ceramic-ahash/src/recon/mod.rs

```diff
@@ -0,0 +1,255 @@
+        }
+        if self.keys.is_empty() {
+            return Message::<H>::default();
+        }
+        let mut received_keys = received.keys.iter();
+        let mut received_hashs = received.ahashs.iter();
+
+        let mut response = Message::<H>::default(); // init the response empty
+
+        let mut left_fencepost = self.keys.first_key_value().unwrap().0;
+        let mut right_fencepost = received_keys
+            .next()
+            .unwrap_or(self.keys.last_key_value().unwrap().0);
+        let mut received_hash = &H::identity();
+        let zero = &H::identity();
+
+        response.keys.push(left_fencepost.to_string());
+        while !received.keys.is_empty() && left_fencepost < received.keys.last().unwrap() {
+            response.process_range(left_fencepost, right_fencepost, received_hash, self);
+            left_fencepost = right_fencepost;
+            right_fencepost = received_keys
+                .next()
+                .unwrap_or(self.keys.last_key_value().unwrap().0);
+            received_hash = received_hashs.next().unwrap_or(zero);
+        }
+        if !received.keys.is_empty() {
+            response.process_range(
+                received.keys.last().unwrap(),
+                self.keys.last_key_value().unwrap().0,
+                zero,
+                self,
+            );
+        }
+        response.end_streak(self.keys.last_key_value().unwrap().0, self);
+        response
+    }
+}
+
+// Messages
+// messages are alternating keys and hashs
+// the hashs are of all keys in range from left to right
+//
+// -> (first, hash(...), middle, hash(...), last)
+// the key to the left of the hash is less then all keys in the hash
+// the key to the right of the hash is grater the all keys in the hash
+// it is like a b-tree node storing a sorted set
+//
+// Represented as a struct of arrays.
+// keys[]
+// hashs[]
+// with one more key then hashes
+#[derive(Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct Message<H: Hash> {
+    #[serde(rename = "k")]
+    pub keys: Vec<String>, // keys must be 1 longer then ahashs unless both are empty
+
+    #[serde(rename = "h")]
+    pub ahashs: Vec<H>, // ahashs must be 1 shorter then keys
+}
+
+impl<H: Hash> Display for Message<H> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "(")?;
+        for (k, h) in self.keys.iter().zip(self.ahashs.iter()) {
+            let hash_hex = if h.is_zero() {
+                "0".to_string()
+            } else {
+                h.to_hex()[0..6].to_string()
+            };
+            write!(f, "{}, {}, ", k, hash_hex)?;
+        }
+
+        write!(f, "{})", self.keys.last().unwrap_or(&"".to_string()))
+    }
+}
+
+pub trait Hash: std::ops::Add<Output = Self> + Clone + Default + PartialEq {
+    fn identity() -> Self {
+        Self::default()
+    }
+    fn is_zero(&self) -> bool;
+    fn clear(&mut self);
+    fn push(&mut self, key: (&String, &AHash));
```

Comment:
Similar question to before, what is the string for? Can we remove it or use a struct to give it self documenting semantics?